### ISSUE TITLE
Default implementation for shouldComponentUpdate?

### ISSUE THREAD

**Issue by fluffynukeit** [paf31/purescript-thermite#10]
[`shouldComponentUpdate`](http://facebook.github.io/react/docs/component-specs.html#updating-shouldcomponentupdate) can be used to examine the incoming properties and state and decide whether the component should be re-rendered.  Given purescript's immutability, we can implement this in the bindings using reference equality similar to this [react mixin](https://github.com/jurassix/react-immutable-render-mixin).

To accommodate edge cases that require mutability, it probably makes sense to include this as part of the `Spec` and add the reference checking as a `Just` default for `simpleSpec`.  Thoughts?

**Comment by paf31** [comment]
Interesting. I wonder if we should use referential equality or an `Eq state` instance. It might cause a slowdown if it's a default implementation, but puts the burden on the developer to implement `Eq`.

So I understand, in what cases does referential equality lead to a performance improvement? For most PS data types, `===` would rarely evaluate to `true` in practice, right?

**Comment by fluffynukeit** [comment]
Just to be clear, it's not just `===` on the objects themselves, but also checking for reference equality on their field values as well.  So in thermite, when properties are assembled into new property objects from the `Props` list, it's true that the reference equality of the property objects will not hold, but it is possible that the reference equality of all their fields holds.  Even if not built from the `Props



A couple cases I encountered while playing with `purescript-signal` and `virtual-dom`.

1.  Let's say you have a model that is processing a fast incoming event stream, like mouse position or drag events, and it's being rendered to a complicated GUI.  The GUI can't render fast enough to keep up, so you decide to downsample using `sampleOn` to sample at, say, 50 ms.  However, `sampleOn` will fire an event every 50 ms whether the signal it's sampling changed or not.  To get around this, I used `distinct'` in combination with `sampleOn`.  The result is a reference equality check every 50 ms, and only when reference equality fails does the GUI re-render.  This puts an upper bound on GUI refresh rate.  Using `shouldComponentUpdate` with reference equality is similar to using `distinct'` in this example.

2. Your GUI displays info about the model state.  Your model processes incoming messages to update state, but some cases results in no change to the state.  In that case, returning the state unchanged to the GUI render function should satisfy reference equality, so the GUI doesn't need to update.  This case is similar to 1) above if the messages are coming in at a high rate.

As for implementation, I think reference equality is the way to go.  My aim here is not to eliminate all unnecessary GUI updates, but only to eliminate those that are easy to guard against in a convenient way.  If the new props and state are the same as the old ones, or their field values are the same as the old values, then there is nothing to update.

Using `Eq` instance, to me, is a little too far removed from the philosophy of "just rerender the GUI every update" because now the `Eq` instance is essentially responsible for determining whether the GUI should redraw or not.  Additionally, a lot
